src/fluree/db/api/query.cljc

```diff
@@ -152,183 +157,110 @@
-                                             (recur r asserted-subjects
-                                                    (update retracted-subjects subject #(vec (conj % flake'))))))))
-                retracted (loop [[subject & r] (vals retracted-subjects)
-                                 acc []]
-                            (if-not subject
-                              acc
-                              (recur r (conj acc (<? (flakes->res db cache fuel 1000000 {:wildcard? true, :select {}}
-                                                                  {} subject))))))
-                asserted  (loop [[subject & r] (vals asserted-subjects)
-                                 acc []]
-                            (if-not subject
-                              acc
-                              (recur r (conj acc (<? (flakes->res db cache fuel 1000000 {:wildcard? true, :select {}}
-                                                                  {} subject))))))]
-            {:block     (:block curr-block)
-             :t         (:t curr-block)
-             :retracted retracted
-             :asserted  asserted})))
-
-
-(defn- format-blocks-resp-pretty
-  [db resp]
-  (async/go-loop [fuel (volatile! 0)
-                  cache (volatile! {})
-                  curr-block (first resp)
-                  rest-blocks (rest resp)
-                  acc []]
-    (let [curr-block' (<? (format-block-resp-pretty db curr-block cache fuel))
-          acc'        (concat acc [curr-block'])]
-      (if (first rest-blocks)
-        (recur fuel cache (first rest-blocks) (rest rest-blocks) acc')
-        acc'))))
-
+(defn history-flakes->json-ld
+  [db q flakes]
+  (go-try
+    (let [fuel    (volatile! 0)
+          cache   (volatile! {})
+          compact (json-ld/compact-fn (fql-parse/parse-context q db))]
+      (->> (sort-by flake/t flakes)
+           (partition-by flake/t)
+           (map (fn [t-flakes]
+                  (go-try
+                    (let [assert-flakes (not-empty (filter flake/op t-flakes))
+                          retract-flakes (not-empty (filter (complement flake/op) t-flakes))
+
+                          asserts (<? (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                                {:wildcard? true, :depth 0} 0 assert-flakes))
+                          retracts (<? (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                                 {:wildcard? true, :depth 0}
+                                                                 0 retract-flakes))]
+
+                      ;; t is always positive for users
+                      (cond-> {:t (- (flake/t (first t-flakes)))}
+                        asserts (assoc :assert asserts)
+                        retracts (assoc :retract retracts))))))
+           (async/merge)
+           (async/into [])
+           (<?)
```

Comment:
the `<?` operator doesn't work with `async/merge`, `async/into`, or the other higher order core async sequence like functions. It only works when you treat channels like promises, not if you are using them as sequential pipeline of asynchronous operations. 

If you're going to use channels as pipelines like this, then you need to isolate the parts of the pipeline that can throw exceptions, wrap those in `try*`/`catch*`, and put those on an error channel that you listen to at the top level with `async/alt!`. You can look at the query pipeline for an example of this.

Here is [where we listen for errors on the error channel](https://github.com/fluree/db/blob/main/src/fluree/db/query/exec.cljc#L44-L60).

Here is [where we catch potential errors and put them on the error channel](https://github.com/fluree/db/blob/main/src/fluree/db/query/exec/where.cljc#L53-L55)

Reply:
Good note! I've got that change ready to go, but while you're looking at this bit of code do you have any ideas on how to preserve the order of the t-flake partitions - right now the `merge` just takes the t-flakes in the order they finish, which shuffles them up. I had an extra sort in there but I think that's pretty wasteful and I suspect there's a way to get it to the results in order - I'm just not familiar enough with all the core.async functions.

Reply:
I suppose we could just put the results into a set and let the consumer deal with it, but I think they'd just sort it and we already have it sorted anyways.

Reply:
Merge doesn't care about order beyond order of completion. If you want to guarantee any other ordering, then you need to put the items who's order you'd like to preserve on a channel, in that order, and then process them sequentially. 

Here is [an example of preserving order in union query patterns in where clauses](https://github.com/fluree/db/blob/main/src/fluree/db/query/exec/where.cljc#L289-L300).

Reply:
Thanks for your help, it's doing the right thing now :D